File: backend/auth-service/src/main/java/com/fixme/authservice/model/ServiceRequest.java
package com.fixme.authservice.model;

import jakarta.persistence.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "service_requests")
public class ServiceRequest {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // Provider (nullable until assigned)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "provider_id", nullable = true)
    private User provider;

    // Customer owner
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id", nullable = false)
    private User customer;

    // Vehicle
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "vehicle_id", nullable = false)
    private Vehicle vehicle;

    @Column(nullable = false, length = 1000)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private VehicleCategory vehicleCategory = VehicleCategory.ALL;

    // ✅ NEW: type of service (OIL_CHANGE / TOWING / etc.)
    @Enumerated(EnumType.STRING)
    @Column(name = "service_type", nullable = false)
    private ServiceType serviceType;

    @Column(nullable = false)
    private Double latitude;

    @Column(nullable = false)
    private Double longitude;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RequestStatus status = RequestStatus.PENDING;

    @Enumerated(EnumType.STRING)
    @Column(name = "progress_stage", nullable = false)
    private ProgressStage progressStage = ProgressStage.ON_THE_WAY;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    public void prePersist() {
        this.createdAt = LocalDateTime.now();
        if (this.progressStage == null) this.progressStage = ProgressStage.ON_THE_WAY;
        if (this.status == null) this.status = RequestStatus.PENDING;
    }

    // ========================
    // Getters / Setters
    // ========================

    public Long getId() {
        return id;
    }

    public User getProvider() {
        return provider;
    }

    public void setProvider(User provider) {
        this.provider = provider;
    }

    public User getCustomer() {
        return customer;
    }

    public void setCustomer(User customer) {
        this.customer = customer;
    }

    public Vehicle getVehicle() {
        return vehicle;
    }

    public void setVehicle(Vehicle vehicle) {
        this.vehicle = vehicle;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public VehicleCategory getVehicleCategory() {
        return vehicleCategory;
    }

    public void setVehicleCategory(VehicleCategory vehicleCategory) {
        this.vehicleCategory = vehicleCategory;
    }

    // ✅ NEW getters/setters
    public ServiceType getServiceType() {
        return serviceType;
    }

    public void setServiceType(ServiceType serviceType) {
        this.serviceType = serviceType;
    }

    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }

    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }

    public RequestStatus getStatus() {
        return status;
    }

    public void setStatus(RequestStatus status) {
        this.status = status;
    }

    public ProgressStage getProgressStage() {
        return progressStage;
    }

    public void setProgressStage(ProgressStage progressStage) {
        this.progressStage = progressStage;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }
}

File: frontend/fixme_app/lib/models/service_request.dart
class ServiceRequest {
  final int id;

  final int customerId;
  final int? providerId;

  final int vehicleId;
  final String plateNumber;
  final String make;
  final String model;
  final int? year;

  final String description;
  final String vehicleCategory;
  final String serviceType;

  final double latitude;
  final double longitude;

  final String status;          // e.g. WAITING_CUSTOMER, ACCEPTED ...
  final String progressStage;   // e.g. ON_THE_WAY, DIAGNOSING, FIXING, DONE
  final String createdAt;

  ServiceRequest({
    required this.id,
    required this.customerId,
    required this.providerId,
    required this.vehicleId,
    required this.plateNumber,
    required this.make,
    required this.model,
    required this.year,
    required this.description,
    required this.vehicleCategory,
    required this.serviceType,
    required this.latitude,
    required this.longitude,
    required this.status,
    required this.progressStage,
    required this.createdAt,
  });

  factory ServiceRequest.fromJson(Map<String, dynamic> j) {
    // progressStage ممكن يوصل null لو DB قديم أو الباكند ما رجّعه
    final ps = (j['progressStage'] ?? j['stage'] ?? 'ON_THE_WAY').toString();

    return ServiceRequest(
      id: (j['id'] as num).toInt(),

      customerId: (j['customerId'] as num).toInt(),
      providerId: (j['providerId'] == null) ? null : (j['providerId'] as num).toInt(),

      vehicleId: (j['vehicleId'] as num).toInt(),
      plateNumber: (j['plateNumber'] ?? '').toString(),
      make: (j['make'] ?? '').toString(),
      model: (j['model'] ?? '').toString(),
      year: (j['year'] == null) ? null : (j['year'] as num).toInt(),

      description: (j['description'] ?? '').toString(),
      vehicleCategory: (j['vehicleCategory'] ?? '').toString(),
      serviceType: (j['serviceType'] ?? '').toString(),
      latitude: (j['latitude'] as num).toDouble(),
      longitude: (j['longitude'] as num).toDouble(), 

      status: (j['status'] ?? '').toString(),
      progressStage: ps,
      createdAt: (j['createdAt'] ?? '').toString(),
    );
  }

  ServiceRequest copyWith({
    String? status,
    String? progressStage,
  }) {
    return ServiceRequest(
      id: id,
      customerId: customerId,
      providerId: providerId,
      vehicleId: vehicleId,
      plateNumber: plateNumber,
      make: make,
      model: model,
      year: year,
      description: description,
      vehicleCategory: vehicleCategory,
      serviceType:serviceType,
      latitude: latitude,
      longitude: longitude,
      status: status ?? this.status,
      progressStage: progressStage ?? this.progressStage,
      createdAt: createdAt,
    );
  }
}
